### ISSUE TITLE
Statistical Data Analysis

### ISSUE THREAD

**Issue by thelasthermit** [thelasthermit/Statistics#1]
#Loading all the necessary libraries for the question
library(ISLR)
library(glmnet)
library(faraway)
library(dplyr)
library(leaps)
library(pls)
library(Metrics)
library(caret)
library(MASS)
library(boot)
library(tree)
library(randomForest)
library(gbm)
library(DT)
library(pls)

#Loading the data into an dataframe object
house.data <- data.frame(read.csv("housingData.csv"), header = T)

#Checking the number of rows and columns of the object
dim(house.data)

summary(house.data)

#From the summary we find that only the 1st column i.e. LotFrontage has missing values:- 259

#Taking a sneakpeak at the data using head()
head(house.data, 10)

#We observe the following:
#A few rows are non-numeric
#A few rows have categorical data
#There is an NA data in the LotFrontage column
#We didn't have the description of NA in the LotFrontage column description
#This needs further exploration

missing.lotfrontage <- sum(is.na(house.data$LotFrontage))
#We have 259 missing data in the LotFrontage column
missing.per <- missing.lotfrontage / nrow(house.data) * 100
missing.per

#18% of the data is missing in the LotFrontage column which is acceptable
#We need to fill this data up as this is a numeric column but with what?

summary(house.data$LotFrontage)
313+21/2

#The mean and the median are really close to the lower bound
#Let's find out the correlation of LotFrontage with other predictors

cor(house.data$LotFrontage, house.data$LotArea, use = "complete.obs")
#The LotArea doesnt have strong correlation with Lot Frontage. 

#Another category is the LotShape which might effect the LotFrontage
summary(house.data$LotFrontage)

house.data.filling <- house.data[,1:3]
group.house.data.filling <- group_by(house.data.filling, LotShape)
groupwise.avg <- data.frame(summarise(group.house.data.filling, LotFrontage.mean =  mean(LotFrontage, na.rm = T )))
groupwise.avg

shape <- house.data$LotShape[10]
shape
if (groupwise.avg[4,1] == shape){
print("yp")
}
groupwise.avg[4,2]

#Now we will replace the NA with the mean found above for each LotShape
for(i in 1:nrow(house.data))
{
	if(is.na(house.data$LotFrontage[i]) == TRUE)
		{
		 shape <- house.data$LotShape[i]
		 for(j in 1:4)
			{
			 if(groupwise.avg[j,1] == shape)
				{
				 house.data$LotFrontage[i] = groupwise.avg[j,2]
				}
			}
		}
}
summary(house.data$LotFrontage)
#We can observe that the NA values have been replaced by the new values
#The average has changed from 69.47 to 70.15

#Our data cleaning part is complete. 
names(house.data)
lm.mod <- lm(house.data$SalePrice~., data = house.data)
summary(lm.mod)
par(mfrow = c(2,2))
plot(lm.mod, pch = 19)

#A long tail is observed in the Q-Q plot.Thus there is a skew in the response variable

hist(house.data$SalePrice, freq = F)
lines(density(house.data$SalePrice))

#As expected there is a left skew in the SlaePrice data due to which there was a long tail

#Using the boxcox method to remove the skewness of the data
bc <- boxcox(lm.mod, lambda = seq(-3,3))

head(house.data$SalePrice, 10)

#The 95% value is really close to 0 thus we can directly go for log transformation of response variable
house.data$SalePrice <- log(house.data$SalePrice)
hist(house.data$SalePrice, freq = F)
lines(density(house.data$SalePrice))

head(house.data$SalePrice, 10)

#Let us again try the OLS LR abd see if the results got better
lm.mod.log <- lm(house.data$SalePrice~., data = house.data)
summary(lm.mod.log)
par(mfrow = c(2,2))
plot(lm.mod)		 

#The trnaform did help in increasing the R-squared but it certainly didn't help much with the Q-Q plot
#We must remove the outliers from the data for better results
#We will remove the points which are more than 3 SDs away from the mean

saleprice.upper <- mean(house.data$SalePrice) + 3*sd(house.data$SalePrice)
saleprice.upper
saleprice.lower <- mean(house.data$SalePrice) - 3*sd(house.data$SalePrice)
saleprice.lower
c = 0
for(i in 1:nrow(house.data))
{
	if((house.data$SalePrice[i] > saleprice.upper)|(house.data$SalePrice[i] < saleprice.lower))
		{
		 c = c + 1
		}
}
paste("There are", c, "outliers in the data")

7/nrow(house.data) * 100
#Only 0.487% of the data would be lost which is not a big loss. 
#We'll remove these 7 points from the data for better results

house.data.test <- house.data
head(house.data.test, 10)
subset(house.data.test, house.data.test$SalePrice > saleprice.lower | house.data.test$SalePrice < saleprice.upper)

for(i in 1:nrow(house.data.test))
{
	if((house.data.test$SalePrice[i] > saleprice.upper)|(house.data.test$SalePrice[i] < saleprice.lower))
	{
		 house.data.test = house.data.test[-i, ]
	}
}
dim(house.data.test)
house.data <- house.data.test
dim(house.data)
house.data.test <- house.data

lm.mod.no.out <- lm(house.data$SalePrice~., data = house.data)
summary(lm.mod.no.out)
par(mfrow = c(2,2))
plot(lm.mod.no.out)

#We'll do the same for the TotalBsmtSF and GrLivArea columns
bsmt.upper <- mean(house.data$TotalBsmtSF) + 3*sd(house.data$TotalBsmtSF)
bsmt.upper
bsmt.lower <- mean(house.data$TotalBsmtSF) - 3*sd(house.data$TotalBsmtSF)
bsmt.lower
dim(house.data.test)
for(i in 1:nrow(house.data.test))
{
	if(house.data.test$TotalBsmtSF[i] > bsmt.upper)
	{
	 	house.data.test = house.data.test[-i, ]
	}
}
dim(house.data.test)
house.data <- house.data.test

grliv.upper <- mean(house.data$GrLivArea) + 3*sd(house.data$GrLivArea)
grliv.upper
grliv.lower <- mean(house.data$GrLivArea) - 3*sd(house.data$GrLivArea)
grliv.lower

for(i in 1:nrow(house.data.test))
{
	if((house.data.test$GrLivArea[i] > grliv.upper)|(house.data.test$GrLivArea[i] < grliv.lower))
	{
	 house.data.test = house.data.test[-i, ]
	}
}
dim(house.data.test)
house.data <- house.data.test
house.data.test <- house.data
dim(house.data)

lm.mod.no.out.2 <- lm(house.data$SalePrice~., data = house.data)
summary(lm.mod.no.out.2)
par(mfrow = c(2,2))
plot(lm.mod.no.out.2)

#We lost less than 2% of the data
#The OLS LR diagnostic plots look a lot better than original data diagnostic plots

#We have two columns with 
house.data[,"Age"] <- house.data[ ,"YrSold"] - house.data[ ,"YearBuilt"]
plot(house.data$Age, house.data$SalePrice)
cor(house.data$Age, house.data$SalePrice)

#We can observe that the categories in the Neighbourhood is 28
#This will cause trouble when we try using regression techniques which converts the dataframe into matrix
#The model matrix would exponentially increase the number of columns
#This will lead to high processing time and might cause error in regression

#We will try to compress the data into 4 categories using 2 criteria
#We will use the Overall Quality and Overall Condition as our criteria
#We will create a new column called "Location"

house.data$Location <- house.data$OverallQual/mean(house.data$OverallQual) + house.data$OverallCond/mean(house.data$OverallCond)
summary(house.data$Location)

chapri.location <- house.data[house.data$Location < 1.8844, "Neighborhood"]
summary(chapri.location)
head(chapri.location)
ameer.location <-house.data[house.data$Location > 2.2140, "Neighborhood"]
summary(ameer.location)

plot(house.data$Location, house.data$SalePrice, pch = 19)
cor(house.data$Location, house.data$SalePrice)

#We can see a correlation between Location and SalePrice columns but it's not strong
#We will now look for alternatives to 28 types of neighborhoods
#We will group each neighborhood and find the average of the Location for each

names(house.data)

group.hood.location <- house.data[, c(5,8,9,29)]
group.hood.location
group.hood.location.categorized <- group_by(group.hood.location, Neighborhood) 
group.hood.avg <- data.frame(summarise(group.hood.location.categorized, Location.mean =  mean(Location)))
mean(group.hood.avg$Location.mean)

uptown <- group.hood.avg[group.hood.avg$Location.mean > 2.01611, ]
chapri <- group.hood.avg[group.hood.avg$Location.mean < 2.01611, ]

mean(uptown$Location.mean)

rich.uptown <- uptown[uptown$Location.mean > 2.135218, ]
dim(rich.uptown)
rich.uptown
norm.uptown <- uptown[uptown$Location.mean < 2.135218, ]
norm.uptown

mean(chapri$Location.mean)

ameer.chapri <- chapri[chapri$Location.mean > 1.906164, ]
ameer.chapri
gareeb.chapri <- chapri[chapri$Location.mean < 1.906164, ]
gareeb.chapri

#We now have 4 categories of the Location criteria
#We will replace the values in the Location column with the following:
#Replace with 1 for the rich.uptown column matching in house.data
#Replace with 2 for the norm.uptown column matching in house.data
#Replace with 3 for the ameer.chapri column matching in house.data
#Replace with 4 for the gareeb.chapri column matching in house.data

house.data.location <- house.data

for(char in rich.uptown$Neighborhood)
{
 house.data.location[house.data.location$Neighborhood == char, "Location"] = 1
}
dim(house.data.location[house.data.location$Location == 1, ])

for(char in norm.uptown$Neighborhood)
{
 house.data.location[house.data.location$Neighborhood == char, "Location"] = 2
}
dim(house.data.location[house.data.location$Location == 2, ])

for(char in ameer.chapri$Neighborhood)
{
 house.data.location[house.data.location$Neighborhood == char, "Location"] = 3
}
dim(house.data.location[house.data.location$Location == 3, ])

for(char in gareeb.chapri$Neighborhood)
{
 house.data.location[house.data.location$Neighborhood == char, "Location"] = 4
}
dim(house.data.location[house.data.location$Location == 4, ])

head(house.data.location[, c(5,29)],10)

house.data <- house.data.location
dim(house.data)
#Now we can use the Location column instead of the Neighborhood column
plot(house.data$Location, house.data$SalePrice)
cor(house.data$Location, house.data$SalePrice)
#We can see a negative correlation between the two
#Since there are a few non-numeric columns thus we'll be unable to use pairs()
#We will separate the numeric columns and try to plot

num.cols <- colnames(house.data[sapply(house.data, is.numeric) == T])
pairs(house.data[num.cols])
cor(house.data[num.cols])
#This worked well and now we have the correlation table for all the predictors

#Since we have already worked so much on the predcitor "Location", we will get rid of the "Neighborhood" predictor
house.data.no.neigh <- house.data[ ,-5]
names(house.data.no.neigh)
#We will now split the data into two parts.
#The first part would be the training set with years <= 2009
#The second part would be the test set with year 2010

pre.house.data <- house.data.no.neigh[house.data.no.neigh$YrSold <= 2009, ]
post.house.data <- house.data.no.neigh[house.data.no.neigh$YrSold == 2010, ]

#Now we'll carry out our regression with the train set and predict the SalePrice for the test set

#We're now all set for out regression analysis

#We'll again split the dataset into training and test datasets

set.seed(1)
var <- sample(nrow(pre.house.data), nrow(pre.house.data)*0.75)

train <- pre.house.data[var, ]
test <- pre.house.data[-var, ]
dim(pre.house.data)
dim(train)
names(train)
unique(train$GarageType)

#We'll start with the Principal Component Regression (PCR)
train.trial <- train

#We'll change the NAs in the GarageType and BsmtFinType1 to "No Garage" and "No Basement" respectively

train.trial[which(is.na(train.trial$GarageType)), ]$GarageType = "No Garage"
unique(train.trial$GarageType)

train.trial[which(is.na(train.trial$BsmtFinType1)), ]$BsmtFinType1 = "No Basement"
unique(train.trial$BsmtFinType1)

train <- train.trial
train.trial$Location
train$Location <- train.trial$Location
train.2 <- data.frame(model.matrix(SalePrice~., data = train)[ ,-1])
train.2$SalePrice = train$SalePrice

test[which(is.na(test$GarageType)), ]$GarageType = "No Garage"
unique(test$GarageType)

test[which(is.na(test$BsmtFinType1)), ]$BsmtFinType1 = "No Basement"
unique(test$BsmtFinType1)
test.2 <- test
test.2 <- data.frame(model.matrix(SalePrice~., data = test)[ ,-1])
test.2$SalePrice = test$SalePrice
test <- test.2
train <- train.2
names(train)
#We'll centre the values of the columns with the numeric values
for(i in length(num.cols))
{
	if((i != 16)|(i!=68))
	{
	 train[ ,num.cols[i]] <- train[ ,num.cols[i]] - mean(train[ ,num.cols[i]])
	}
}

pca.train <- prcomp(train[ ,-69], center = T)
summary(pca.train)
pca.train$rot
par(mfrow = c(1,1))
biplot(pca.train, choices = c(1,2))

pcr.mod <- pcr(exp(SalePrice)~., data = train, validation = "CV", ncomp = 50)
summary(pcr.mod)
validationplot(pcr.mod, val.type = "MSEP")
try(fit.pcr <- predict(pcr.mod, newdata = test, ncomp = 21),silent = T) 
?try
fit.pcr
mean((exp(test$SalePrice) - fit.pcr)^2)
R2.pcr <- cor(fit.pcr, exp(train$SalePrice))^2
R2.pcr
predict(pcr.mod, test)
mse.pcr <- mse(test$SalePrice, predict(pcr.mod, data = test, ncomp = 10))
mse.pcr

#We chose 21 PCs model 
#We'll  now go for the Bagging technique

#############################Bagging Technique##################################

bag.mod <- randomForest(exp(train$SalePrice)~., data = train, mtry = ncol(train)-1, importance = T, ntree = 1000)
plot(bag.mod)
importance(bag.mod)
varImpPlot(bag.mod)
fit.bag <- predict(bag.mod, newdata = test)
fit.bag
mse.bag <- mse(exp(test$SalePrice), fit.bag)
mse.bag
mean((exp(test$SalePrice) - fit.bag)^2)

#We'll now use the randomForest method for prediction
rf.mod <- randomForest(exp(train$SalePrice)~., data = train, mtry = (ncol(train)-1)/3, importance = T, ntree = 600)
plot(rf.mod)
mse.rf <- mse(test$SalePrice, predict(rf.mod, data = test))
mse.rf
importance(rf.mod)
varImpPlot(rf.mod)

#We'll now use the Boosting technique for Regression
boost.mod <- gbm(exp(train$SalePrice)~., data = train, distribution = "gaussian", n.trees = 1500, interaction.depth = 2, shrinkage = 0.003, verbose = F)
boost.mod
which.min(boost.mod$cv.error)
summary(boost.mod)
par(mfrow = c(1,2))
plot(boost.mod, i = "OverallQual")
plot(boost.mod, i = "Location")
fit.boost <- predict(boost.mod, newdata = test)
mse.boost <- mse(exp(test$SalePrice), fit.boost)
mse.boost

#We'll now choose the best of all the methods (Lowest RMSE)

#Creating a dataframe object to store the values of the RMSEs of all the models above
MSE <- data.frame(c("Bagging", "Random Forest", "Boosting", "Principle Component Regression"), NA)
colnames(MSE)[1] = "Model"
colnames(MSE)[2] = "RMSE"
MSE
MSE[1,2] <- mse.bag
MSE[2,2] <- mse.rf
MSE[3,2] <- mse.boost
MSE[4,2] <- mse.pcr
MSE

#From the table it is evident that the best model is the Boosting
#It has the lowest MSE ~0.228

#Using that method we'll try to predict the prices of the houses sold in 2010
#Our dataset is the post.house.data whose SalePrice has to be predicted

post.house.data.2 <- post.house.data

post.house.data.2[which(is.na(post.house.data.2$GarageType)), ]$GarageType = "No Garage"
unique(post.house.data.2$GarageType)
post.house.data.2[which(is.na(post.house.data.2$BsmtFinType1)), ]$BsmtFinType1 = "No Basement"
unique(post.house.data.2$BsmtFinType1)
test.post2009 <- data.frame(model.matrix(SalePrice~., data = post.house.data.2)[ ,-1])
test.post2009$SalePrice <- post.house.data$SalePrice

mse.rf.post <- mse(test.post2009$SalePrice, predict(rf.mod, data = test.post2009))
mse.rf.post
mse.bag.post <- mse(test.post2009$SalePrice, predict(bag.mod, data = test.post2009))
mse.bag.post
mse.boost.post <- mse(test.post2009$SalePrice, predict(boost.mod, data = test.post2009))
mse.boost.post
